components/pkg-export-helm/src/chart.rs

```diff
@@ -0,0 +1,202 @@
+use std::io;
+use std::io::Write;
+
+use common::ui::{UI, Status};
+use export_docker;
+use export_docker::Result;
+use export_k8s::{Error, Manifest, ManifestJson};
+
+use chartfile::ChartFile;
+use values::Values;
+
+pub struct Chart<'a> {
+    name: String,
+    habitat_name: String,
+    chartfile: ChartFile,
+    manifest_template: ManifestJson,
+    values: Values,
+    ui: &'a mut UI,
+}
+
+impl<'a> Chart<'a> {
+    pub fn new_for_cli_matches(ui: &'a mut UI, matches: &clap::ArgMatches) -> Result<Self> {
+        if !matches.is_present("NO_DOCKER_IMAGE") {
+            export_docker::export_for_cli_matches(ui, &matches)?;
+        }
+        let manifest = Manifest::new_from_cli_matches(ui, &matches)?;
+        let name = matches
+            .value_of("CHART")
+            .unwrap_or(&manifest.habitat_name)
+            .to_string();
+        let version = matches.value_of("VERSION");
+        let description = matches.value_of("DESCRIPTION");
+        let chartfile = ChartFile::new(&name, version, description)?;
+
+        Chart::new_for_manifest(manifest, name, chartfile, ui)
+    }
+
+    fn new_for_manifest(
+        manifest: Manifest,
+        name: String,
+        chartfile: ChartFile,
+        ui: &'a mut UI,
+    ) -> Result<Self> {
+        let main = json!({
+            "metadata_name": "{{.Values.metadataName}}",
+            "habitat_name": "{{.Values.habitatName}}",
+            "image": "{{.Values.imageName}}",
+            "count": "{{.Values.instanceCount}}",
+            "service_topology": "{{.Values.serviceTopology}}",
+            "service_group": manifest.service_group.clone().map(|_| "{{.Values.serviceGroup}}"),
+            "config_secret_name": manifest.config_secret_name
+                .clone()
+                .map(|_| "{{.Values.configSecretName}}"),
+                "ring_secret_name": manifest.ring_secret_name
+                .clone()
+                .map(|_| "{{.Values.ringSecretName}}"),
+                "bind": manifest.binds,
+        });
+
+        let mut values = Values::new();
+        values.add_entry("metadataName", &manifest.metadata_name);
+        values.add_entry("habitatName", &manifest.habitat_name);
+        values.add_entry("imageName", &manifest.image);
+        values.add_entry("instanceCount", &manifest.count.to_string());
+        values.add_entry("serviceTopology", &manifest.service_topology.to_string());
+        if let Some(ref group) = manifest.service_group {
+            values.add_entry("serviceGroup", group);
+        }
+        if let Some(ref name) = manifest.config_secret_name {
+            values.add_entry("configSecretName", name);
+        }
+        if let Some(ref name) = manifest.ring_secret_name {
+            values.add_entry("ringSecretName", name);
+        }
+
+        let mut binds = Vec::new();
+        let mut i = 0;
+        for bind in &manifest.binds {
+            let split: Vec<&str> = bind.split(":").collect();
+            if split.len() < 3 {
```

Comment:
Would be really great to have binds to be guaranteed to be correct by the manifest itself. It's probably second or third time I see the bind string splitting code. That would also make this function to return `Self` instead of `Result<Self>`…

Reply:
Good point. I did think about it: https://github.com/habitat-sh/habitat/blob/master/components/pkg-export-kubernetes/src/manifest.rs#L47 . Perhaps I should address this as part of this work.